exec/java-exec/src/main/java/org/apache/drill/exec/physical/impl/join/JoinTemplate.java

```diff
@@ -87,125 +42,68 @@ public void setupJoin(FragmentContext context, JoinStatus status, VectorContaine
    */
   public final boolean doJoin(final JoinStatus status) {
     while(!status.isOutgoingBatchFull()) {
-      // for each record
 
-      // validate input iterators (advancing to the next record batch if necessary)
-      if (!status.isRightPositionAllowed()) {
+      if (status.rightFinished()) {
         if (((MergeJoinPOP)status.outputBatch.getPopConfig()).getJoinType() == JoinRelType.LEFT) {
```

Comment:
let's grab this property at the beginning of the join rather than each time we doJoin.